text/0300-programmatic-toolkit.md

```diff
@@ -0,0 +1,745 @@
+# Programmatic Toolkit
+
+* **Original Author(s):**: @mrgrain
+* **Tracking Issue**: #300
+* **API Bar Raiser**: @rix0rrr
+
+The [Programmatic Toolkit] allows integrators to build lifecycle management solutions for CDK applications.
+It provides them with full control over all output and allows to interweave actions with custom steps.
+
+## Working Backwards
+
+### README
+
+[Programmatic Toolkit] is a library to run CDK actions programmatically.
+The package provides implementations for the following actions:
+
+- `bootstrap`
+- `list`
+- `synth`
+- `deploy`
+- ...
+
+#### Self-contained example
+
+```ts
+// Create a [Programmatic Toolkit]
+await using cdk = new Toolkit({
+  ioHost: new CloudWatchLogsIoHost('my-log-stream'), // write all output to CW logs
+  // awsSdkProvider: new DefaultAwsSdk(), // optional
+});
+
+// Define the CloudExecutable
+const cx = CloudExecutable.fromCloudAssemblyProducer(async (context) => {
+  const app = new cdk.App({ context });
+  const stack = new cdk.Stack(app);
+  return app.synth();
+});
+
+// Synth and store a cached CloudExecutable, this won't run synth again
+const cxCache = await cdk.synth(cx);
+
+// Deploy our pre-synth'ed app
+const deployment = await cdk.deploy(cxCache, { stacks: ['MyTestStack'] });
+
+// Returns a list of ARNs of the deployment
+return deployment.listResources().filter(r => r.arn);
+```
+
+#### Toolkit setup
+
+A default toolkit can be setup with no additional configuration:
+
+```ts
+const cdk = new Toolkit()
+```
+
+You can customize the functionality of the toolkit.
+The main options are provided through `ioHost` and `awsSdkProvider`.
+
+```ts
+const cdk = new Toolkit({
+  ioHost: new MyIoHost(),
+  awsSdkProvider: new MyAwsSdkProvider(),
+})
+```
+
+#### Messages and Requests
+
+The toolkit emits *messages* and *requests* to structure interactions.
+A *request* is a *message* that requires the receiver to respond, otherwise the toolkit will continue with a default response.
+
+For every emitted message, the toolkit calls the `notify` or `requestResponse` method of an `IoHost` (more on this below).
+Every call is `await`ed for its completion, allowing to receiver to perform asynchronous operations like HTTP calls.
+It's okay for the receiver not do anything with the input and return immediately.
+The type definition of messages and requests is as follows:
+
+```ts
+interface IoMessage<T> {
```

Comment:
I think this needs to be a tagged union.

```typescript
type IoMessage = 
  | { code: 'deploy' } & TypedIoMessage<{ arn: string }>
  | { code 'message' } & TypedIoMessage<string>
  | ...
```

Reply:
Explained. Will make a note.